[Do Not Merge] Add Information Protection APIs

ranisha2 opened the pull request:
Add Information Protection APIs:
- GET Informaton Protection Policy
- PUT Information Protetion Policy
- LIST Information Protection Policies

APIs are not public yet. We need to merge this only after exposing the APIs in production.

This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process.

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [x] If applicable, the PR references the bug/issue that it fixes.
- [x] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [x] My spec meets the review criteria:
  - [x] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [x] The spec follows the guidelines described in the [Swagger checklist](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md) document.
  - [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

ranisha2 committed:
Add Information Protection APIs

Add Information Protection APIs:
- GET Informaton Protection Policy
- PUT Information Protetion Policy
- LIST Information Protection Policies

AutorestCI commented:
# Automation for azure-sdk-for-go

AutorestCI commented:
# Automation for azure-sdk-for-python

AutorestCI commented:
# Automation for azure-sdk-for-node

AutorestCI commented:
# Automation for azure-libraries-for-java

devigned commented on specification/securitycenter/resource-manager/Microsoft.Security/preview/2017-08-01-preview/informationProtection.json:
```diff
@@ -0,0 +1,310 @@
+                        "$ref": "#/definitions/SensitivityLabel"
+                    }
+                },
+                "sensitivityInformationTypes": {
```
Please don't make this an array. This should be a dictionary of `key`: `SensitivityLabel`. I say this because it seems you have already created an order field within the `SensitivityLabel` type, which seems to obviate the need for an array type.

Also, by making this a dictionary, the service could better support idempotent PATCH requests for individual items rather than having to PUT the entire array each time.

ranisha2 replied:
We changed to dictionary.

devigned commented on specification/securitycenter/resource-manager/Microsoft.Security/preview/2017-08-01-preview/informationProtection.json:
```diff
@@ -0,0 +1,310 @@
+                    "readOnly": true,
+                    "description": "Describes the last UTC time the policy was modified"
+                },
+                "sensitivityLabels": {
```
Same feedback as https://github.com/Azure/azure-rest-api-specs/pull/3119/files#r191520615. Please change to a dictionary rather than an array.

ranisha2 replied:
Done

devigned commented on specification/securitycenter/resource-manager/Microsoft.Security/preview/2017-08-01-preview/examples/InformationProtectionPolicies/CreateOrUpdateInformationProtectionPolicy_example.json:
```diff
@@ -0,0 +1,238 @@
+            "properties": {
+                "sensitivityLabels": [
+                    {
+                        "id": "1345da73-bc5a-4a8f-b7dd-3820eb713da8",
```
How would these GUIDs get generated in a ARM template?

ranisha2 replied:
This is a string, user can set any id.
Our built-in labels have guids as id.
We'll update our examples to make it clear.

ranisha2 replied:
Actually we decided that they should be guids, but these are determined by users and it can be any guid they choose.

devigned commented on specification/securitycenter/resource-manager/Microsoft.Security/preview/2017-08-01-preview/examples/InformationProtectionPolicies/CreateOrUpdateInformationProtectionPolicy_example.json:
```diff
@@ -0,0 +1,238 @@
+                        "id": "7fb9419d-2473-4ad8-8e11-b25cc8cf6a07",
+                        "name": "Networking",
+                        "order": 100,
+                        "recommendedLabelId": "575739d2-3d53-4df0-9042-4c7772d5c7b1"
```
How would `recommendedLabelId` be generated in an ARM template? Wouldn't this become a non-idempotent PUT in an ARM template?

ranisha2 replied:
This is not generated, it should reference the corresponding recommended label.
User will set and ID of a specific label, then he'll reuse that id here

devigned commented on specification/securitycenter/resource-manager/Microsoft.Security/preview/2017-08-01-preview/informationProtection.json:
```diff
@@ -0,0 +1,310 @@
+                    "type": "number",
+                    "description": "The order of the information type."
+                },
+                "recommendedLabelId": {
```
Does `recommendedLabelId` have to be a GUID, or can it simply be a string? If it must be a GUID, then it should not be of type string, but typed as a uuid.

ranisha2 replied:
No, it represents a label ID, and it can be a string.

ranisha2 replied:
Actually we decided that they should be guids. Will update the type

ravbhatnagar replied:
Why does it need to be a GUID? Are you using it for some kind of uniqueness? GUIDs hamper the usability considerably.

ranisha2 replied:
We're planning to integrate with Microsoft Information Protection and Azure Information Protection in the future, and they work with guids. Also built-in information types have pre-defined guids and they are globally unique.

devigned commented on specification/securitycenter/resource-manager/Microsoft.Security/preview/2017-08-01-preview/informationProtection.json:
```diff
@@ -0,0 +1,310 @@
+                    "type": "string",
+                    "description": "The name of the information type."
+                },
+                "order": {
```
Does order have a limit? Why does order need to exist if the data type is within an array?

ranisha2 replied:
We changed it to a dictionary

ranisha2 commented on specification/securitycenter/resource-manager/Microsoft.Security/preview/2017-08-01-preview/informationProtection.json:
```diff
@@ -0,0 +1,310 @@
+            "type": "object",
+            "description": "The information type.",
+            "properties": {
+                "id": {
```
No, it can be a string

ranisha2 replied:
Actually we decided that they should be guids. Will update the type.

ranisha2 commented on specification/securitycenter/resource-manager/Microsoft.Security/preview/2017-08-01-preview/informationProtection.json:
```diff
@@ -0,0 +1,310 @@
+                    }
+                ],
+                "responses": {
+                    "200": {
```
It's not a long running operation.
Added response code 201 (Should support both)

ranisha2 replied:
It's not a long running operation. Also the policy always exists (Default policy), so PUT requests will always update the existing one.

ranisha2 commented:
Will update the PR soon.

ranisha2 committed:
Update information protection swagger

Update information protection swagger

AutorestCI commented:
# Automation for azure-sdk-for-ruby

ranisha2 committed:
Fix build error

Fix build error - Add missing property "items"

ravbhatnagar commented on specification/securitycenter/resource-manager/Microsoft.Security/preview/2017-08-01-preview/informationProtection.json:
```diff
@@ -0,0 +1,360 @@
+        }
+    },
+    "paths": {
+        "/{scope}/providers/Microsoft.Security/informationProtectionPolicies/{informationProtectionPolicyName}": {
```
Is this a tracked or proxy resource?

ranisha2 replied:
It's a proxy resource

ravbhatnagar commented on specification/securitycenter/resource-manager/Microsoft.Security/preview/2017-08-01-preview/informationProtection.json:
```diff
@@ -0,0 +1,360 @@
+            }
+        }
+    },
+    "paths": {
```
Is there a DELETE support for this?

ranisha2 replied:
We do not support DELETE for now, as from the perspective of the user, the policy always exists (Cannot delete the built-in policy)

ranisha2 committed:
Information Protection API fixes

- Update keywords to be an array of a keyword object.
- Fix "enabled" property type in both InformationType and SensitivityLabel

dsgouda commented:
@ravbhatnagar Thanks
@ranisha2 adding my review soon

dsgouda commented on specification/securitycenter/resource-manager/Microsoft.Security/preview/2017-08-01-preview/informationProtection.json:
```diff
@@ -0,0 +1,362 @@
+                            "$ref": "#/definitions/ErrorResponse"
+                        }
+                    }
+                }
```
Please check if this is a long running operation

ranisha2 replied:
It's not a long running operation

dsgouda commented on specification/securitycenter/resource-manager/Microsoft.Security/preview/2017-08-01-preview/informationProtection.json:
```diff
@@ -0,0 +1,362 @@
+            "type": "object",
+            "description": "describes properties of an information protection policy.",
+            "properties": {
+                "lastModifiedUtc": {
```
Please check if any of these properties are read only

ranisha2 replied:
This is the only one

dsgouda commented on specification/securitycenter/resource-manager/readme.md:
````diff
@@ -0,0 +1,59 @@
+
+``` yaml
+openapi-type: arm
+tag: package-composite-v1
````
Tags should usually correspond to the api version which is `2017-08-01-preview` in this case

dsgouda commented on specification/securitycenter/resource-manager/readme.md:
````diff
@@ -0,0 +1,59 @@
+
+``` yaml $(csharp)
+csharp:
+  azure-arm: true
````
remove this line

ranisha2 commented on specification/securitycenter/resource-manager/Microsoft.Security/preview/2017-08-01-preview/informationProtection.json:
```diff
@@ -0,0 +1,362 @@
+                "description": "Details of the information protection policy.",
+                "operationId": "InformationProtectionPolicies_Get",
+                "parameters":
+                [{
```
Fixed

ranisha2 committed:
InformationProtection API fixes

Remove custom/builtin properties and add a boolean "custom" property to the information type itself.

AutorestCI commented:
# Automation for azure-sdk-for-java

dsgouda commented:
@ranisha2 This is good to go if you can address the tag naming

ranisha2 committed:
Merge pull request #1 from Azure/master

Sync

ranisha2 committed:
Merge with Microsoft.Security

Move to security folder to merge with Microsoft.Security

dsgouda commented:
@ranisha2 This looks good. Let me know when this is good to merge and I shall do so

chlahav commented on specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/informationProtection.json:
```diff
@@ -0,0 +1,351 @@
+{
+    "swagger": "2.0",
```
collapse to the main file
we still have changes that we are doing and we would like to have central control over them
this is reason for all my comments

chlahav commented on specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/informationProtection.json:
```diff
@@ -0,0 +1,351 @@
+{
+    "swagger": "2.0",
+    "info": {
+        "title": "ASC API",
```
change title to security center

dsgouda commented:
@ranisha2 please address the comments, will close the PR if there is not activity by next week

dsgouda commented:
@ranisha2 closing this due to inactivity.
Please open a new PR referencing this when you have addressed pending comments and ready to merge
